src/shogun/clustering/KMeansBase.cpp

```diff
@@ -130,6 +131,115 @@ void KMeansBase::compute_cluster_variances()
 	}
 }
 
+SGMatrix<float64_t> KMeansBase::compute_std_dev() const
+{
+    require(cluster_centers.size() > 0, "KMeans is not trained!");
+
+    auto lhs_cache = distance->get_lhs();
+    distance->replace_lhs(distance->get_rhs());
+
+	SGMatrix<float64_t> stds(dimensions, k);
+
+	auto rhs = distance->get_rhs()->as<DenseFeatures<float64_t>>();
+	SGVector<int32_t> cluster_assignments;
+	SGVector<int64_t> weights_set;
+	std::tie(cluster_assignments, weights_set, std::ignore) =
+	    compute_cluster_assignments(k);
+
+	auto cluster_indexes = std::vector<SGVector<index_t>>(k);
+	auto cluster_counters = SGVector<index_t>(k);
+	for (int32_t current_cluster = 0; current_cluster < k; ++current_cluster)
+	{
+		cluster_indexes[current_cluster] =
+		    SGVector<index_t>(weights_set[current_cluster]);
+		cluster_counters[current_cluster] = 0;
+	}
+
+	for (int32_t i = 0; i < rhs->get_num_vectors(); ++i)
+	{
+		int point_cluster = cluster_assignments[i];
+		int& cluster_counter = cluster_counters[point_cluster];
+		cluster_indexes[point_cluster][cluster_counter] = i;
+		++cluster_counter;
+	}
+
+	for (int32_t i = 0; i < k; ++i)
+	{
+		stds.set_column(i, view(rhs, cluster_indexes[i])->std());
+	}
+
+    distance->replace_lhs(lhs_cache);
+	return stds;
+}
+
+std::tuple<SGVector<int32_t>, SGVector<int64_t>, int32_t>
+KMeansBase::compute_cluster_assignments(
+    int32_t num_centers,
+    std::function<void(ChangeCentersContext)> change_centers,
+    SGVector<int32_t> cluster_assignments,
+    SGVector<int64_t> weights_set) const
+{
+    auto lhs = distance->get_lhs()->as<DenseFeatures<float64_t>>();
+
+    int32_t lhs_size=lhs->get_num_vectors();
+    int32_t dim=lhs->get_num_features();
+
+    if (!cluster_assignments.size()) {
+        cluster_assignments = SGVector<int32_t>(lhs_size);
+    }
+
+    if (!weights_set.size())
+    {
+        /* Weights : Number of points in each cluster */
+        weights_set = SGVector<int64_t> (num_centers);
+        weights_set.zero();
+        /* Initially set all weights for zeroth cluster, Changes in assignement step */
+        weights_set[0] = lhs_size;
+    }
+
+    int32_t changed = 0;
+#pragma omp parallel for firstprivate(lhs_size, dim, num_centers) \
+		shared(cluster_assignments, weights_set) \
+		reduction(+:changed) if (!change_centers)
+    /* Assigment step : Assign each point to nearest cluster */
+    for (int32_t i=0; i<lhs_size; i++)
+    {
+        const int32_t cluster_assignments_i=cluster_assignments[i];
+        int32_t min_cluster, j;
+        float64_t min_dist, dist;
```

Comment:
could you do this in two lines and init them with the values below straight away? less error prone